### PR TITLE
[CDAP-20947] Skip enabling metadata consumer subscriber if no consumers are enabled

### PR DESCRIPTION
**Pull request by itsankit-google**
JIRA: [CDAP-20947](https://cdap.atlassian.net/browse/CDAP-20947)

Tested:
```
2024-01-24 18:30:02,976 - INFO [main:i.c.c.m.e.k.MetadataServiceMain@124] - Skipping enabling MetadataConsumerSubscriberService, no metadata consumer extensions are enabled.
2024-01-24 18:30:02,985 - INFO  [main:i.c.c.m.e.k.AbstractServiceMain@217] - Service io.cdap.cdap.master.environment.k8s.MetadataServiceMain initialized
2024-01-24 18:30:02,994 - INFO  [main:i.c.c.m.e.k.AbstractServiceMain@222] - Starting all services for io.cdap.cdap.master.environment.k8s.MetadataServiceMain
2024-01-24 18:30:02,997 - INFO  [main:i.c.c.m.e.k.AbstractServiceMain@224] - Starting service MessagingMetricsCollectionService [NEW] for io.cdap.cdap.master.environment.k8s.MetadataServiceMain
2024-01-24 18:30:03,009 - INFO  [main:i.c.c.m.e.k.AbstractServiceMain@224] - Starting service MetadataService [NEW] for io.cdap.cdap.master.environment.k8s.MetadataServiceMain
```

[CDAP-20947]: https://cdap.atlassian.net/browse/CDAP-20947?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/metadata/MetadataConsumerSubscriberService.java [hunk 1]
```diff
@@ -178,6 +183,10 @@ protected void doStartUp() throws Exception {
   protected void processMessages(StructuredTableContext structuredTableContext,
                                  Iterator<ImmutablePair<String, MetadataMessage>> messages)
     throws IOException, ConflictException {
+    if (!isEnabled) {
+      LOG.info("Skipping processing messages, no consumers are enabled.");
```
Comment (tivv) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/metadata/MetadataConsumerSubscriberService.java]: Please say "extensions" here as the main consumer will still be running in the other service. Also consider adding it to the variable name.

**Review by tivv — CHANGES_REQUESTED** [review]
This will not help much, or for our bug even would make this worse as consumer would still be running and reading messages, just doing "NOP" on processing. If you really want to skip processing, you need to skip installing this service.

Also please add JIRA to the PR title

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/metadata/MetadataConsumerSubscriberService.java (hunk 1 above)
Comment (itsankit-google) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/metadata/MetadataConsumerSubscriberService.java]: ack

File: cdap-master/src/main/java/io/cdap/cdap/master/environment/k8s/MetadataServiceMain.java [hunk 2]
```diff
@@ -113,7 +116,11 @@ protected void addServices(Injector injector, List<? super Service> services,
       EnvironmentOptions options) {
     services.add(injector.getInstance(MetadataService.class));
     services.add(injector.getInstance(MetadataSubscriberService.class));
-    services.add(injector.getInstance(MetadataConsumerSubscriberService.class));
+    Collection<String> enabledMetadataConsumers = cConf.getStringCollection(
+        Constants.MetadataConsumer.METADATA_CONSUMER_EXTENSIONS_ENABLED_LIST);
+    if (enabledMetadataConsumers != null && !enabledMetadataConsumers.isEmpty()) {
+      services.add(injector.getInstance(MetadataConsumerSubscriberService.class));
+    }
```
Comment (tivv) [inline cdap-master/src/main/java/io/cdap/cdap/master/environment/k8s/MetadataServiceMain.java:128]: Could you please log something if we don't add this service?

Comment (itsankit-google) [reply cdap-master/src/main/java/io/cdap/cdap/master/environment/k8s/MetadataServiceMain.java:128]: Done.

File: cdap-master/src/main/java/io/cdap/cdap/master/environment/k8s/MetadataServiceMain.java [hunk 3]
```diff
@@ -113,7 +118,15 @@ protected void addServices(Injector injector, List<? super Service> services,
       EnvironmentOptions options) {
     services.add(injector.getInstance(MetadataService.class));
     services.add(injector.getInstance(MetadataSubscriberService.class));
-    services.add(injector.getInstance(MetadataConsumerSubscriberService.class));
+
+    if (cConf != null && cConf.getStringCollection(
+        Constants.MetadataConsumer.METADATA_CONSUMER_EXTENSIONS_ENABLED_LIST).isEmpty()) {
+      LOG.debug("Skipping enabling MetadataConsumerSubscriberService, "
```
Comment (tivv) [inline cdap-master/src/main/java/io/cdap/cdap/master/environment/k8s/MetadataServiceMain.java]: Make it info. It's one line on the start of the service and it's important

**Review by tivv — APPROVED** [review]
Please also fix javadoc

File: cdap-master/src/main/java/io/cdap/cdap/master/environment/k8s/MetadataServiceMain.java (hunk 3 above)
Comment (itsankit-google) [reply cdap-master/src/main/java/io/cdap/cdap/master/environment/k8s/MetadataServiceMain.java]: Ack, done.

**Review by tivv — APPROVED** [review]

**Pull request merged**